java/ql/lib/semmle/code/java/security/OverlyLargeRangeQuery.qll

```diff
@@ -0,0 +1,282 @@
+  private string highCut(string cut) {
+    cut = ["Z", "z", "9"] and
+    result = cut
+    or
+    cut = ["A", "a", "0"] and
+    next(result) = cut
+  }
+
+  /** Gets the cutoff char used for a given `part` of a range when pretty-printing it. */
+  private string cutoff(OverlyWideRange range, int part) {
+    exists(int low, int high | isRange(range, low, high) |
+      result =
+        rank[part + 1](string cut |
+          cut = cutoffs() and low < toCodePoint(cut) and toCodePoint(cut) < high
+        |
+          cut order by toCodePoint(cut)
+        )
+    )
+  }
+
+  /** Gets the number of parts we should print for a given `range`. */
+  private int parts(OverlyWideRange range) { result = 1 + strictcount(cutoff(range, _)) }
+
+  /** Holds if the given part of a range should span from `low` to `high`. */
+  private predicate part(OverlyWideRange range, int part, string low, string high) {
+    // first part.
+    part = 0 and
+    (
+      range.isRange(low, high) and
+      parts(range) = 1
+      or
+      parts(range) >= 2 and
+      range.isRange(low, _) and
+      high = highCut(cutoff(range, part))
+    )
+    or
+    // middle
+    part >= 1 and
+    part < parts(range) - 1 and
+    low = lowCut(cutoff(range, part - 1)) and
+    high = highCut(cutoff(range, part))
+    or
+    // last.
+    part = parts(range) - 1 and
+    low = lowCut(cutoff(range, part - 1)) and
+    range.isRange(_, high)
+  }
+
+  /** Gets an escaped `char` for use in a character class. */
+  bindingset[char]
+  private string escape(string char) {
+    exists(string reg | reg = "(\\[|\\]|\\\\|-|/)" |
+      if char.regexpMatch(reg) then result = "\\" + char else result = char
+    )
+  }
+
+  /** Gets a part of the equivalent range. */
+  private string printEquivalentCharClass(OverlyWideRange range, int part) {
+    exists(string low, string high | part(range, part, low, high) |
+      if
+        isAlphanumeric(low) and
+        isAlphanumeric(high)
+      then result = low + "-" + high
+      else
+        result =
+          strictconcat(string char | char = getInRange(low, high) | escape(char) order by char)
+    )
+  }
+
+  /** Gets the entire pretty printed equivalent range. */
+  string printEquivalentCharClass(OverlyWideRange range) {
+    result =
+      strictconcat(string r, int part |
+        r = "[" and part = -1 and exists(range)
+        or
+        r = printEquivalentCharClass(range, part)
+        or
+        r = "]" and part = parts(range)
+      |
+        r order by part
+      )
+  }
+}
+
+/** Gets a char range that is overly large because of `reason`. */
+RegExpCharacterRange getABadRange(string reason, int priority) {
+  priority = 0 and
+  reason = "is equivalent to " + result.(OverlyWideRange).printEquivalent()
+  or
+  priority = 1 and
+  exists(RegExpCharacterRange other |
+    reason = "overlaps with " + other + " in the same character class" and
+    result != other and
```

Comment:
Superfluous, as it has now been moved into `overlap`. (Also, I'd hope that the `rankRange` condition below implies this.)
```suggestion
```